PayloadPkg/OsLoader/PreOsChecker.c

```diff
@@ -0,0 +1,149 @@
+} PRE_OS_PAYLOAD_PARAM;
+
+typedef VOID (*PRE_OS_CHECKER_ENTRY) (VOID *Params);
+
+STATIC VOID *mPreOsCheckerEntry = NULL;
+
+/**
+  To see whether PreOsChecker is loaded properly
+
+  @retval           TRUE if successfully loaded, otherwise FALSE
+
+**/
+BOOLEAN
+EFIAPI
+IsPreOsCheckerLoaded (
+  IN  VOID
+  )
+{
+  return ((mPreOsCheckerEntry == NULL) ? FALSE : TRUE);
+}
+
+/**
+  Search for and load pre-OS checker for execution.
+
+  This function will search for a pre-OS checker binary that might
+  be loaded as part of the payload, if found get the entry point for
+  execution later instead of jumping into the OS directly.
+
+  @retval  EFI_LOAD_ERROR         Pre-OS checker is not loaded successfully.
+  @retval  EFI_INVALID_PARAMETER  Invalid parameter is used at loading PreOsChecker image
+  @retval  EFI_UNSUPPORTED        Pre-OS checker image is not supported
+
+**/
+EFI_STATUS
+EFIAPI
+LoadPreOsChecker (
+  IN  VOID
+  )
+{
+  EFI_STATUS         Status;
+  UINT32             Length;
+  VOID              *Buffer;
+
+  //
+  // TBD: Set to NULL to re-load PreOsChecker in case of restarting Payload for now.
+  //
+  mPreOsCheckerEntry = NULL;
+  //if (mPreOsCheckerEntry != NULL) {
+  //  DEBUG ((DEBUG_INFO, "Pre-OS checker is already loaded at entry @ 0x%p\n", mPreOsCheckerEntry));
+  //  return EFI_STATUS;
+  //}
+
+  //
+  // Check if there is a pre-OS checker that needs to be executed
+  //
+  Buffer = NULL;
+  Length = 0;
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_POSC, &Buffer, &Length);
+
+  if (EFI_ERROR (Status) || (Buffer == NULL) || (Length == 0)) {
+    return EFI_NOT_FOUND;
+  }
+
+  Status = LoadElfImage (Buffer, &mPreOsCheckerEntry);
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Pre-OS checker entry @ 0x%p\n", mPreOsCheckerEntry));
+  }
+
+  return Status;
+}
+
+/**
+  Start PreOsChecker with OS information
+
+  @param[in] OsBootParam          Pointer of OS BOOT_PARAMS
+
+  @retval EFI_NOT_FOUND           PreOsChecker not found
+  @retval EFI_INVALID_PARAMETER   Invalid OS BOOT_PARAMS found
+  @retval none                    No EFI_SUCCESS if PreOsChecker is executed properly
+
+**/
+EFI_STATUS
+EFIAPI
+StartPreOsChecker (
+  IN  BOOT_PARAMS   *OsBootParam
+  )
+{
+  PRE_OS_PAYLOAD_PARAM       PreOsParams;
+  PRE_OS_CHECKER_ENTRY       EntryPoint;
+
+  if (mPreOsCheckerEntry == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  if (OsBootParam == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  PreOsParams.Version     = 0x1;
+  PreOsParams.HeapSize    = EFI_SIZE_TO_PAGES (0);
+  PreOsParams.HeapAddr    = (UINT32) AllocatePages (PreOsParams.HeapSize);
+  PreOsParams.HobListPtr  = PcdGet32 (PcdPayloadHobList);
+
+  PreOsParams.OsBootState.Esi     = (UINT32) OsBootParam;
+  PreOsParams.OsBootState.Eip     = OsBootParam->Hdr.Code32Start;
+  PreOsParams.OsBootState.Eflags  = 0;
+
+  EntryPoint = (PRE_OS_CHECKER_ENTRY)mPreOsCheckerEntry;
+  EntryPoint (&PreOsParams);
+
+  //
+  // Not reach here
```

Comment:
Should we add a CPU dead loop or ASSERT here for debug?

Reply:
Updated in coming push. Thanks.